### PR TITLE
Add kubeconfig delete support to kubeform.

### PR DESCRIPTION
**Pull request by madhusudancs**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1426)

### CODE REVIEW

File: kubeform/pkg/provider/provider.go [hunk 1]
```diff
@@ -166,7 +161,24 @@ func createKubeconfig(d *schema.ResourceData, meta interface{}) error {
...
+
+	po := clientcmd.NewDefaultPathOptions()
+	if path, ok := d.GetOk("path"); ok {
+		po.LoadingRules.ExplicitPath = path.(string)
+	}
+
+	// Retry until modifyConfig succeeds or times out.
+	log.Printf("[DEBUG] updating kubeconfig")
+	if !poll(cfg.configPollInterval, cfg.ConfigPollTimeout, updateConfig(po, cfg.kubeConfig, removeConfig)) {
+		return fmt.Errorf("cluster components never turned healthy")
```
Comment (errordeveloper) [inline kubeform/pkg/provider/provider.go]: I am not entierely sure about why health check is needed here, if we are to delete it. One of the reason for deleting something may well be that it's broken.

Comment (errordeveloper) [inline kubeform/pkg/provider/provider.go]: Can this error be somehow ignore upstream?

**Comment by errordeveloper** [comment]
In order to review this, I'd like to test it myself, especially because I'd like to really understand the usage.

File: kubeform/pkg/provider/provider.go (hunk 1 above)
Comment (madhusudancs) [inline kubeform/pkg/provider/provider.go]: This is just a case of error string copy/paste fail. I fixed this in the `createKubeconfig()` function which you reviewed earlier today and overlooked it here.

We do not check for cluster's health here, just for kubeconfig to update, i.e. delete the entry. If the operation to delete the kubeconfig entries fail, it must be reported to the user, so this is important.

**Comment by madhusudancs** [comment]
> In order to review this, I'd like to test it myself, especially because I'd like to really understand the usage.

@errordeveloper Sure, please do give it a try and let me know what you think.

File: kubeform/pkg/provider/provider.go (hunk 1 above)
Comment (madhusudancs) [inline kubeform/pkg/provider/provider.go]: I have fixed the error message now. Thanks for catching it!

**Comment by madhusudancs** [comment]
@errordeveloper Forgot to mention that there is a Makefile now. Building the code should be as simple as running `make build`. Also, here is a sample of how I am using it https://github.com/madhusudancs/kubernetes-anywhere/blob/federation/phase1/gce/gce.jsonnet#L201

**Comment by errordeveloper** [comment]
> Also, here is a sample of how I am using it...

Ok, great, that's what I need to see to see the most!

File: kubeform/pkg/provider/provider.go [hunk 2]
```diff
@@ -304,9 +316,23 @@ func kubeConfigGetter(d *schema.ResourceData) clientcmd.KubeconfigGetter {
...
-func updateConfig(configAccess clientcmd.ConfigAccess, suppliedConfig *clientcmdapi.Config) func() (bool, error) {
+func kubeconfigInit(d *schema.ResourceData, meta interface{}) (*config, error) {
+	configF := meta.(configFunc)
+	cfg, err := configF(d)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the cluster client: %v", err)
+	}
+
+	if len(cfg.kubeConfig.Clusters) != 1 || len(cfg.kubeConfig.AuthInfos) != 1 || len(cfg.kubeConfig.Contexts) != 1 {
+		return nil, fmt.Errorf("config must supplied for exactly one cluster - number of clusters: %d, number of users: %d, number of contexts: %d", len(cfg.kubeConfig.Clusters), len(cfg.kubeConfig.AuthInfos), len(cfg.kubeConfig.Contexts))
```
Comment (errordeveloper) [inline kubeform/pkg/provider/provider.go]: must **be**

File: kubeform/pkg/provider/provider.go [hunk 3]
```diff
@@ -304,9 +316,23 @@ func kubeConfigGetter(d *schema.ResourceData) clientcmd.KubeconfigGetter {
 	}
 }
 
-func updateConfig(configAccess clientcmd.ConfigAccess, suppliedConfig *clientcmdapi.Config) func() (bool, error) {
+func kubeconfigInit(d *schema.ResourceData, meta interface{}) (*config, error) {
+	configF := meta.(configFunc)
+	cfg, err := configF(d)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the cluster client: %v", err)
```
Comment (errordeveloper) [inline kubeform/pkg/provider/provider.go]: perhaps this should say "Kubernetes" instead of "cluster"

**Comment by test-foxish** [comment]
recomputing cla status...

**Comment by errordeveloper** [comment]
@madhusudancs are planning to continue working on this?

**Comment by grodrigues3** [comment]
[APPROVALNOTIFIER] The Following OWNERS Files Need Approval:

- **kubeform**

We suggest the following people:
cc @lavalamp 
 You can indicate your approval by writing `/approve` in a comment
 You can cancel your approval by writing `/approve cancel`in a comment

**Pull request closed without merging**